### PR TITLE
Create JavaType to marshal JNI types to JS.

### PR DESCRIPTION
**Pull request by szurbrigg**
- Update JavaMethod to use instances of JavaType for args and return values.
- Organize C++ source files into directories.

### CODE REVIEW

File: duktape/src/main/jni/java/JavaMethod.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    auto parameterType = env->GetObjectArrayElement(parameterTypes, i);
+    m_argumentLoaders[i]  = typeMap.get(env, static_cast<jclass>(parameterType));
+  }
+
+  const jmethodID getReturnType =
+      env->GetMethodID(methodClass, "getReturnType", "()Ljava/lang/Class;");
+  jobject returnTypeObject = env->CallObjectMethod(method, getReturnType);
+
+  const jmethodID methodId = env->FromReflectedMethod(method);
+  const auto returnType = typeMap.get(env, static_cast<jclass>(returnTypeObject));
```
Comment (swankjesse) [inline duktape/src/main/jni/java/JavaMethod.cpp]: supernit: does this belong immediately below line 40 ?

File: duktape/src/main/jni/java/JavaMethod.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  std::vector<jvalue> args(m_argumentLoaders.size());
+  // Load the arguments off the stack and convert to Java types.
+  // Note we're going backwards since the last argument is at the top of the stack.
+  for (ssize_t i = m_argumentLoaders.size() - 1; i >= 0; --i) {
+    args[i] = m_argumentLoaders[i]->pop(ctx, env);
+  }
+
+  return m_methodBody(ctx, env, javaThis, args.data());
```
Comment (swankjesse) [inline duktape/src/main/jni/java/JavaMethod.cpp]: cute

File: duktape/src/main/jni/java/JavaType.cpp [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+  virtual const char* getBoxMethodName() const { return "valueOf"; }
+
+  bool isPrimitive() const override { return true; }
+  jclass boxedClass() const { return static_cast<jclass>(m_boxedClassRef.get()); }
+
+private:
+  const GlobalRef m_boxedClassRef;
+};
+
+class BoxedPrimitive : public JavaType {
```
Comment (swankjesse) [inline duktape/src/main/jni/java/JavaType.cpp]: Is it possible to define this after the unboxed primitives?

Comment (swankjesse) [inline duktape/src/main/jni/java/JavaType.cpp]: (as is I was a bit surprised when `Boolean` is an unboxed primitive)

File: duktape/src/main/jni/java/JavaType.cpp [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+} // anonymous namespace
+
+JavaTypeMap::~JavaTypeMap() {
+  for (auto entry : m_types) {
+    delete entry.second;
+  }
+}
+
+const JavaType* JavaTypeMap::get(JNIEnv* env, jclass c) {
+  if (m_types.empty()) {
```
Comment (swankjesse) [inline duktape/src/main/jni/java/JavaType.cpp]: cute

**Comment by swankjesse** [comment]
LGTM!

File: duktape/src/main/jni/java/JavaMethod.cpp (hunk 1 above)
Comment (szurbrigg) [inline duktape/src/main/jni/java/JavaMethod.cpp]: Moved!

File: duktape/src/main/jni/java/JavaType.cpp (hunk 3 above)
Comment (szurbrigg) [inline duktape/src/main/jni/java/JavaType.cpp]: Sure thing!

**Pull request merged**